### ISSUE TITLE
How to Inject and extract context for distributed tracing for asynchronous tasks running on different process?

### ISSUE THREAD

**Issue by davidlinowolt** [DataDog/dd-trace-java#5535]
Hi!, I'm looking for help debugging a Datadog trace setup.

Here's my case:

I have a HTTP request that goes to Service X which after processing, it schedules a task by writing the task to the Database.

Then, on another process, there's a watcher that queries the database constantly and picks up the task and process it.

The watcher code is something like this:
```kotlin
 while (true) {
            task = findTaskInDB()
            if (task != null) {
                    processTask(task)
            } else {
                    sleep(TIME)
            }
        }
```

I'm trying to link those 2 processes into the same trace.

The closest to a similar setup I found on DD dog was [this](https://docs.datadoghq.com/tracing/trace_collection/opentracing/java/#inject-and-extract-context-for-distributed-tracing) using OpenTracing libs.

I partially succeeded!

My approach was to, on the Http request handling side, extract the tracer context and add it to the task:
```kotlin
@Serializable
data class SpanContextJson(val fields: Map<String, String>)
override fun addTracingContextTo(task: TracedTask) {
        val tracer = GlobalTracer.get()
        val spanContextMap = mutableMapOf<String, String>()
        tracer.inject(tracer.activeSpan().context(), Format.Builtin.TEXT_MAP, TextMapAdapter(spanContextMap))

        val parentSpanContext = SpanContextJson(spanContextMap)
        task.tracingContext = Json.encodeToString(SpanContextJson.serializer(), parentSpanContext)
    }
    
```
Then on the task worker, I get that context from the DB, create a new span and activate it:

```kotlin
fun addTracingContextTo(task: TracedTask) {
        val tracer = GlobalTracer.get()
        val spanContextMap = mutableMapOf<String, String>()
        // Extract the span context from the global tracer
        tracer.inject(tracer.activeSpan().context(), Format.Builtin.TEXT_MAP, TextMapAdapter(spanContextMap))

        val parentSpanContext = SpanContextJson(spanContextMap)

        // Inject the JSON string representation of the span context
        task.tracingContext = Json.encodeToString(SpanContextJson.serializer(), parentSpanContext)
    }

    class ChildSpan(operationName: String, parentContext: SpanContext) : Closeable {
        private val span: Span

        init {
            val tracer = GlobalTracer.get()
            span = tracer.buildSpan(operationName).asChildOf(parentContext).start()
            tracer.activateSpan(span)
        }

        override fun close() {
            span.finish()
        }
    }
```

So now, the worker code looks like this:

```kotlin
 while (true) {
            task = findTaskInDB()
            if (task != null) {
                    distributedTracer.traceOperationWithContextFrom("taskWatcher", task).use {
                             processTask(task)
                    }
            } else {
                    sleep(TIME)
            }
        }
```    
Now both traces are linked! Yay!

The problem is that, even though I close the span, Datadog keeps adding the periodical DB queries to check for a new task as part of the same trace.

When a new task is picked up, then a new span is created and the previous one is closed. 
So, my guess is that there's some leak with the Tracer and/or some unexpected side-effect.

What am I missing?

**Comment by davidlinowolt** [comment]
And I think the span * is * closed.
But the x-datadog-trace-id  lingers and affects subsequent spans. Until that is replaced by a new task (edited) 

In the docs for correlating logs and traces manually, they remove it from Log4j's thread context: https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/java/?tab=log4j2#manual-injection

I think I need to do something equivalent, but for the tracer....

**Comment by davidlinowolt** [comment]
This is how my flame graph looks like: 
![image](https://github.com/DataDog/dd-trace-java/assets/135004951/df5ceb6c-aabf-42f3-9499-65b2d6c7767b)

And the purple lines keep going until a new job is picked up

**Comment by davidlinowolt** [comment]
This is less than ideal,
But I sorted this out by wrapping `findTaskInDB()` in a new span that ignores the active context:

```
val span = tracer.buildSpan("findTaskInDB").ignoreActiveSpan().start()
try (Scope scope = tracer.activateSpan(span)) {
    task = findTaskInDB()
} finally {
    span.finish()
}
```

**Comment by PerfectSlayer** [comment]
Hi @davidlinowolt 

This is kind of hard to follow what's the current state of the code and how it behaves but I can see from the `ChildSpan` class, you are creating a scope by activating a span you don't close.

**Comment by davidlinowolt** [comment]
@PerfectSlayer the span is closed.
The `use` calls `close()` at the end of the block, and `close` finished the span

**Comment by dougqh** [comment]
Tracer.activateSpan returns a Scope object that also needs to be closed to deactivate the span.
As long as the span is active on the local thread, any new spans or logging will be associated with it -- even if the span is finished.

**Comment by PerfectSlayer** [comment]
Here is a sample to illustrate the missing scope closing logic:

```kotlin
class ChildSpan(operationName: String, parentContext: SpanContext) : Closeable {
        private val span: Span
        private val scope: Scope

        init {
            val tracer = GlobalTracer.get()
            span = tracer.buildSpan(operationName).asChildOf(parentContext).start()
            scope = tracer.activateSpan(span)
        }

        override fun close() {
            scope.close()
            span.finish()
        }
    }
```

Check the scope object instance that seems missing from your sample.

**Comment by davidlinowolt** [comment]
A-ha!

I didn't expect that the scope needed to be closed as well.



@PerfectSlayer thanks for the help. I missed that bit of your original message. My bad.

**Comment by davidlinowolt** [comment]
Tried by closing the scope as well.
Now it works as intended.

**Comment by PerfectSlayer** [comment]
Glad to hear that 🎉
